### PR TITLE
feat: expose _TrainContext.from_config() [DET-4217]

### PR DESCRIPTION
**Pull request by rb-determined-ai**
_TrainContext.from_config() creates a training config via the same
mechanism that we use for local test mode.  In effect, this allows users
to instantiate their own Trials and mess with them them in totally
framework-native ways.

An alternate strategy which was considered was to create some sort of
DummyContext, which would have been a totally transparent object that
took a sample of hparams as its input, and which visibly did not
interfere with any objects (i.e. wrap_model() would just return its
input).

The from_config() strategy was chosen because various context functions
(like custom reducers in estimator, or data_layer caching) are useful
to user model code, but would have to be reimplemented with the
DummyContext.  Or, alternately, the user would have to only use a subset
of context methods when they used the DummyContext.  A real context that
should work with any real Trial seems far more useful, at a slight cost
to transparency.

### CODE REVIEW

File: harness/tests/experiment/test_local.py [hunk 1]
```diff
@@ -13,3 +21,62 @@ def test_test_one_batch() -> None:
...
+    weight = 0.0
+    for _epoch in range(1):
+        for data in range(data_len):
+            grad = trial.calc_gradient(weight, [data])
+            weight -= lr * grad
+
+    # Simluate validation loss.
+    sim_loss = trial.calc_loss(weight, range(data_len))
+
+    assert pytest.approx(sim_loss, eval_loss)
```
Comment (rb-determined-ai) [inline harness/tests/experiment/test_local.py:82]: @davidhershey these tests should hopefully be a starting point for the docs you were talking about

Comment (davidhershey) [reply harness/tests/experiment/test_local.py:82]: I LOVE IT

File: harness/determined/pytorch/_data.py [hunk 2]
```diff
@@ -191,6 +203,18 @@ def get_data_loader(
             worker_init_fn=self.worker_init_fn,  # type: ignore
         )
 
+    def __iter__(self) -> Iterator:
+        """Compatibiliy with the real DataLoader when using a PyTorchTrial outside of Determined."""
+        return iter(self.get_data_loader())
+
+    def __len__(self) -> int:
+        """Compatibiliy with the real DataLoader when using a PyTorchTrial outside of Determined."""
+        # TODO(DET-1524): uncomment this logic.
```
Comment (shiyuann) [inline harness/determined/pytorch/_data.py:212]: can you be more prescriptive on when we should uncomment this logic?

File: harness/determined/pytorch/_pytorch_context.py [hunk 3]
```diff
@@ -360,6 +360,8 @@ def configure_apex_amp(
         return models, optimizers
 
     def _should_communicate_and_update(self) -> bool:
+        if not self.env.training:
```
Comment (shiyuann) [inline harness/determined/pytorch/_pytorch_context.py:363]: Why is this function called in non-training mode?

File: harness/determined/estimator/_estimator_context.py [hunk 4]
```diff
@@ -150,7 +157,7 @@ class EstimatorExperimentalContext(_data_layer.DataLayerContext):
 
     def __init__(self, env: det.EnvContext, hvd_config: horovod.HorovodContext) -> None:
         super().__init__(env=env, hvd_config=hvd_config)
-        self._allgather_fn = None  # type: Optional[Callable[[Any], List]]
+        self._allgather_fn = default_allgather_fn  # type: Callable[[Any], List]
```
Comment (shiyuann) [inline harness/determined/estimator/_estimator_context.py:160]: What is it to fix?

File: harness/determined/_train_context.py [hunk 5]
```diff
@@ -18,6 +18,14 @@ def __init__(self, env: det.EnvContext, hvd_config: horovod.HorovodContext):
         self.distributed = DistributedContext(env, hvd_config)
         self._stop_requested = False
 
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "_TrainContext":
+        env_context, rendezvous_info, hvd_config = det._make_local_execution_env(
+            training=False,
```
Comment (shiyuann) [inline harness/determined/_train_context.py:24]: it looks this feature is used for training. Why is it set to be false?

File: harness/determined/estimator/_estimator_context.py (hunk 4 above)
Comment (rb-determined-ai) [reply harness/determined/estimator/_estimator_context.py:160]: Otherwise you get an error when you use custom reducers outside of determined.  I explained in the docstring of `default_allgather_fn`

File: harness/determined/_train_context.py (hunk 5 above)
Comment (rb-determined-ai) [reply harness/determined/_train_context.py:24]: because `training` really means "training inside determined", and was originally contrasted with "exporting" basically.

In this case, users are using their trial code outside of determined and we should just hand them back their unmodified models and optimizers and such, as with checkpoint export.

Comment (rb-determined-ai) [reply harness/determined/_train_context.py:24]: fun fact, after writing this code I found `det.experimental.create_trial_instance` makes the same exact calls, only it inverts the control that the user would have compared to `_TrainContext.from_config()`.

File: harness/determined/pytorch/_pytorch_context.py (hunk 3 above)
Comment (rb-determined-ai) [reply harness/determined/pytorch/_pytorch_context.py:363]: It's called in `step_optimizer()`, which is called as part of any `train_batch`, regardless of whether or not the code is running in a task container.

File: harness/determined/pytorch/_data.py (hunk 2 above)
Comment (rb-determined-ai) [reply harness/determined/pytorch/_data.py:212]: Sure.

File: harness/determined/_train_context.py [hunk 6]
```diff
@@ -18,6 +18,14 @@ def __init__(self, env: det.EnvContext, hvd_config: horovod.HorovodContext):
         self.distributed = DistributedContext(env, hvd_config)
         self._stop_requested = False
 
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "_TrainContext":
```
Comment (aaron276h) [inline harness/determined/_train_context.py:22]: non-blocking: prolly should add a docstring

**Review by aaron276h — COMMENTED** [review]
This looks real good, what is the plan for documenting this / providing examples how and when to use this feature?

**Comment by rb-determined-ai** [comment]
> This looks real good, what is the plan for documenting this / providing examples how and when to use this feature?

@davidhershey has already volunteered to take on that honorable task : )

File: harness/determined/_train_context.py (hunk 6 above)
Comment (rb-determined-ai) [reply harness/determined/_train_context.py:22]: done

**Review by aaron276h — APPROVED** [review]

File: harness/determined/_train_context.py (hunk 5 above)
Comment (rb-determined-ai) [reply harness/determined/_train_context.py:56]: This brings up a good point though.  We currently have this experimental function that we've never promoted (and which I had totally forgotten about) called `create_trial_instance()`.  The behavior is very similar, but the API has bit of a different feel to it:

```python
# 1. Proposed API
config = { ... }
context = det.pytorch.PyTorchTrialContext.from_config(config)
trial = MyPyTorchTrial(context)
```

vs

```python
# 2. API similar to what we already have
config = { ... }
trial = det.create_trial_instance(MyPyTorchTrial, config)
```

`1` is a bit more verbose but perhaps a bit less magical, since users would actually be instantiating their Trial manually rather than calling a function that does the instantiation automatically.  Not sure what real users would think of that tradeoff, or if it would even matter very much.

@aaron276h @shiyuann @davidhershey, what do you think?

Comment (rb-determined-ai) [reply harness/determined/_train_context.py:56]: I guess there is also a third option:
```python
# 3. Just an idea
config = { ... }
trial = MyPyTorchTrial.from_config(config)
```
But I think it might be a bit confusing to users how we go from a `config` in `MyPyTorchTrial.from_config(config)` to a `context` in `MyPytorchTrial.__init__(context)`

Also, if you allow bare `context` objects via `det.pytorch.PyTorchTrialContext.from_config(config)` you enable users to interactively play with e.g. data layer functions in a notebook without having to go through a Trial class.

Yeah, actually I think I really option `1` because I like exposing the context object and letting users feed it to their own trial.

Comment (aaron276h) [reply harness/determined/_train_context.py:56]: I think option 1 is definitely the best one, its the most clear about what is going on and the extra line users will be required to add isn't a big overhead.

Comment (davidhershey) [reply harness/determined/_train_context.py:56]: I also think 1) is the best option.

**Review by shiyuann — APPROVED** [review]
It would be nice if you could rename `self.training` to be something like `self.in_trial_runner`. But IDK how difficult it would be for you to do that.

**Comment by rb-determined-ai** [comment]
@shiyuann and I talked off-thread and settled on `env_context.managed_training`, since it applies in local test mode (or local training mode if we had it).
